cpp/ql/lib/semmle/code/cpp/TemplateParameter.qll

```diff
@@ -0,0 +1,92 @@
+/**
+ * Provides a hierarchy of classes for modeling C/C++ template parameters.
+ */
+
+import semmle.code.cpp.Type
+private import semmle.code.cpp.internal.ResolveClass
+
+abstract private class TemplateParameterImpl extends Locatable {
+  override string getAPrimaryQlClass() { result = "TemplateParameterImpl" }
+}
+
+/**
+ * A C++ template parameter.
+ *
+ * In the example below, `T`, `TT`, and `I` are template parameters:
+ * ```
+ * template <class T, template<typename> TT, int I>
+ * class C { };
+ * ```
+ */
+final class TemplateParameterBase = TemplateParameterImpl;
+
+/**
+ * A C++ non-type template parameter.
+ *
+ * In the example below, `I` is a non-type template parameter:
+ * ```
+ * template <int I>
+ * class C { };
+ * ```
+ */
+class NonTypeTemplateParameter extends Literal, TemplateParameterImpl {
```

Comment:
I would rename this to `LiteralTemplateParameter`. If there are other types of non-type template parameter we can add them.

Reply:
"non-type" is the name used by the C++ standard, so I think "NonTypeTemplateParameter" is most appropriate.

Reply:
It's a literal sort of by accident, due to the way our frontend chooses to represent these.